File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/verticalTraversal_987.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

import java.util.*;

public class verticalTraversal_987 {
    //      3
    //   1     4
    // 0   2  2
    // [3,1,4,0,2,2]->[[0],[1],[3,2,2],[4]]
    // col row values
    TreeMap<Integer, TreeMap<Integer, PriorityQueue<Integer>>> map;
    List<List<Integer>> res;

    public List<List<Integer>> verticalTraversal(TreeNode root) {
        map = new TreeMap<>();
        res = new ArrayList<>();
        helper(root, 0, 0);
        map.forEach((col, treeMap) -> {
            ArrayList<Integer> tmp = new ArrayList<>();
            treeMap.forEach((row, queue) -> {
                while (!queue.isEmpty()) tmp.add(queue.poll());
            });
            res.add(tmp);
        });
        return res;
    }

    private void helper(TreeNode root, int row, int col) {
        if (root == null) return;
        if (!map.containsKey(col))
            map.put(col, new TreeMap<>());
        if (!map.get(col).containsKey(row))
            map.get(col).put(row, new PriorityQueue<>());
        map.get(col).get(row).add(root.val);
        helper(root.left, row + 1, col - 1);
        helper(root.right, row + 1, col + 1);
    }
}

File: Clion/ZeroTrac/1600_1800/verticalTraversal_987.cpp
//
// Created by TzashiNorpu on 9/30/2022.
//
#include "../header/1600_1800.h"

using namespace ZeroTrac;

vector<vector<int>> Solution::verticalTraversal(TreeNode *root) {
  vector<tuple<int, int, int>> nodes;
  function<void(TreeNode *, int, int)> dfs = [&](TreeNode *node, int row, int col) {
      if (!node) return;
      nodes.emplace_back(col, row, node->val);
      dfs(node->left, row + 1, col - 1);
      dfs(node->right, row + 1, col + 1);
  };
  dfs(root, 0, 0);
  sort(nodes.begin(), nodes.end());
  vector<vector<int>> res;
  int lastCol = INT_MIN;
  for (auto &[col, row, val]: nodes) {
    if (col != lastCol) lastCol = col, res.emplace_back();
    res.back().emplace_back(val);
  }
  return res;
}